spring-cloud-gcp-starters/spring-cloud-gcp-starter-logging/src/main/java/org/springframework/cloud/gcp/logging/autoconfig/StackdriverLoggingAutoConfiguration.java

```diff
@@ -20,24 +20,62 @@
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.logging.CompositeTraceIdExtractor;
 import org.springframework.cloud.gcp.logging.LoggingWebMvcConfigurer;
+import org.springframework.cloud.gcp.logging.TraceIdExtractor;
 import org.springframework.cloud.gcp.logging.TraceIdLoggingWebMvcInterceptor;
+import org.springframework.cloud.gcp.logging.XCloudTraceIdExtractor;
+import org.springframework.cloud.gcp.logging.ZipkinTraceIdExtractor;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Import;
 
 /**
  * @author Mike Eltsufin
+ * @author Chengyuan Zhao
  */
 @Configuration
 @ConditionalOnClass({ TraceIdLoggingWebMvcInterceptor.class, TraceLoggingEnhancer.class })
+@EnableConfigurationProperties({ StackdriverLoggingProperties.class })
 @ConditionalOnProperty(value = "spring.cloud.gcp.logging.enabled", matchIfMissing = true)
 @Import(LoggingWebMvcConfigurer.class)
 public class StackdriverLoggingAutoConfiguration {
 
 	@Bean
 	@ConditionalOnMissingBean
-	public TraceIdLoggingWebMvcInterceptor getLoggingWebMvcInterceptor() {
-		return new TraceIdLoggingWebMvcInterceptor();
+	public TraceIdLoggingWebMvcInterceptor loggingWebMvcInterceptor(
+			TraceIdExtractor extractor) {
+		return new TraceIdLoggingWebMvcInterceptor(extractor);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public TraceIdExtractor traceIdExtractor(
+			StackdriverLoggingProperties loggingProperties) {
+		TraceIdExtractorType checkedType = loggingProperties.getExtractorType();
+		if (checkedType == null) {
+			checkedType = TraceIdExtractorType.XCLOUD_ZIPKIN;
+		}
```

Comment:
How about

```
TraceIdExtractorType checkedType = loggingProperties.getExtractorType() != null
  ? loggingProperties.getExtractorType()
  : TraceIdExtractorType.XCLOUD_ZIPKIN;
```

`getExtractorType()` is O(1), so it should be OK to repeat.

Reply:
Let's not get carried away here. :)
We've already spent a considerable amount of engineering time on this.

Reply:
Just copy and paste it.

Reply:
No, I prefer without ternary operator here. Much cleaner to read and there won't be a [Speculative execution](https://en.wikipedia.org/wiki/Speculative_execution) leading to [Spectre](https://spectreattack.com/) vulnerability. 😄 

Redundant joke here, of course, but no reason to argue any more. The code is good and doesn't repeat method call for nothing.